openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -1,144 +1,203 @@
-                        frame[2][3:])
-                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+    # Holds temporary callstack
+    # List with each element 4-tuple(filename, line number, function name, text)
+    # and filtered with respect to regular expressions
+    temp_call_stack = [(frame[0],
+                        frame[1],
+                        frame[2],
+                        frame[3])
+                       for frame in [frames for frames in traceback.extract_stack()]
                        if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
 
-    # avoid duplication.
-    if temp_call_stack not in STACK_BOOK[current_model] and TRACK_FLAG \
-            and not issubclass(current_model, tuple(HALT_TRACKING)):
-        STACK_BOOK[current_model].append(temp_call_stack)
-        log.info("logging new call stack for %s:\n %s", current_model, temp_call_stack)
+    def _print(frame):
+        return str('\n File ' + str(frame[0]) + ', line number ' + str(frame[1]) + ', in ' +
+                   str(frame[2]) + '\n\t' + str(frame[3]))
+
+    # Customize output of call stack
+    final_call_stack = ""
+    for frame in temp_call_stack:
+        final_call_stack += _print(frame)
+
+    def _should_get_logged(entity_name):
+        if not HALT_TRACKING:
+            if TRACK_FLAG and temp_call_stack not in STACK_BOOK[entity_name]:  # TRACK_FLAG False iff @donottrack
+                return True
+            else:
+                return False
+        else:
+            if inspect.isclass(entity_name):
+                if temp_call_stack not in STACK_BOOK[entity_name] and not issubclass(entity_name, tuple(HALT_TRACKING[-1])):
+                    return True
+                else:
+                    return False
+            else:  # Assumption : Everything other than "class" will be passed as string
+                if temp_call_stack not in STACK_BOOK[entity_name] and entity_name not in tuple(HALT_TRACKING[-1]):
+                    return True
+                else:
+                    return False
+
+    if _should_get_logged(entity_name):
+        STACK_BOOK[entity_name].append(temp_call_stack)
+        log.info("Logging new call stack number %s for %s:\n %s", len(STACK_BOOK[entity_name]),
+                 entity_name, final_call_stack)
 
 
 class CallStackMixin(object):
-    """ A mixin class for getting call stacks when Save() and Delete() methods are called
+    """ Mixin class for getting call stacks when save() and delete() methods are called
     """
 
     def save(self, *args, **kwargs):
         """
-        Logs before save and overrides respective model API save()
+        Logs before save() and overrides respective model API save()
         """
-        capture_call_stack(type(self))
+        if hasattr(self, 'model'):
+            capture_call_stack(self.model)
+        else:
+            capture_call_stack(type(self))
```

Comment:
This should only be mixed into a django model, right? So, `self.model` should always be set.